### PR TITLE
NOTIF-190 consume kafka messages in backend

### PR DESCRIPTION
**Pull request by bennetelli**

### CODE REVIEW

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:258]: This returns a `Uni` with no subscription afterwards which mean the `Uni` may never be resolved (executed). Does it actually send an email?

Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:265]: I don't know. I spend so much time in trying to test it but this still doesn't work.

**Comment by theute** [comment]
Is it a draft or is it ready for review&merge ? If it's still a draft, please mark it as draft so we don't accidentally merge

**Comment by bennetelli** [comment]
> Is it a draft or is it ready for review&merge ? If it's still a draft, please mark it as draft so we don't accidentally merge

I am not able to test the email sending thing. The unit test is always green, even if I remove all the mocks. I'm waiting for Gwennegs response here.

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:265]: [This guide](https://smallrye.io/smallrye-mutiny/guides/testing) contains some information about testing `Uni`/`Multi` but it doesn't really explain how to use `UniAssertSubscriber` so a [quick look](https://github.com/search?l=Java&q=UniAssertSubscriber&type=Code) at the GitHub code from all repos that use `UniAssertSubscriber` can also be helpful. That's one way to test reactive code, but if we strictly consider the question of the emails being sent or not, I don't think `UniAssertSubscriber` is the easiest way to test that.

In notifications-backend, a webhook is called each time an email is sent so mocking the webhook server side with `mockserver` could be a solution to check that the emails are sent. We do that in [WebhookTest](https://github.com/RedHatInsights/notifications-backend/blob/master/backend/src/test/java/com/redhat/cloud/notifications/processors/webhook/WebhookTest.java) for example. But that's a bit overkill because I don't think the entire webhooks code should be tested here.

Mocking the `WebhookTypeProcessor webhookSender` attribute from `EmailSubscriptionTypeProcessor` with Mockito and verifying that the `doHttpRequest` method is called is probably enough to make sure emails are sent during the `EmailSubscriptionTypeProcessor#consumeEmailAggregations` execution.

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:269]: This (untested code written from GH) should fix the issue.

I added a few comments for the review process but don't commit them 😃 
```suggestion
    @Incoming("aggregation")
    // If the method returns Uni<Void> instead of void, Quarkus will automatically subscribe to the Uni.
    public Uni<Void> consumeEmailAggregations(String aggregationsJson) throws JsonProcessingException {
        List<EmailAggregation> aggregations = objectMapper.readValue(aggregationsJson, new TypeReference<>() { });
        // Each element of the "aggregations" list must be processed as a reactive (Uni) operation because the email won't be sent otherwise, so we need to transform the List into a Multi.
        return Multi.createFrom().iterable(aggregations)
                // Each element of the Multi will be treated as a Uni.
                // Concatenate prevents concurrent access to the Hibernate Reactive session (that's forbidden).
                .onItem().transformToUniAndConcatenate(emailAggregationItem -> {
                    Action action = new Action();
                    action.setContext(emailAggregationItem.getPayload().getMap());
                    action.setEvents(List.of());
                    action.setAccountId(emailAggregationItem.getAccountId());
                    action.setApplication(emailAggregationItem.getApplicationName());
                    action.setBundle(emailAggregationItem.getBundleName());

                    // We don't have a eventtype as this aggregates over multiple event types
                    action.setEventType(null);
                    action.setTimestamp(LocalDateTime.now(ZoneOffset.UTC));

                    // We don't have any endpoint (yet) as this aggregates multiple endpoints
                    Notification item = new Notification(action, null);

                    return sendEmail(item, EmailSubscriptionType.DAILY);
                })
                // The following line will wait for all the Unis above to be resolved and then replace everything with a Uni<Void> value because that's all we need in the end.
                .onItem().ignoreAsUni();
    }
```

File: backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java
Comment (gwenneg) [inline backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java:66]: `EmailSubscriptionTypeProcessor#consumeEmailAggregations` should be tested using [SmallRye Reactive Messaging testing features](https://smallrye.io/smallrye-reactive-messaging/smallrye-reactive-messaging/2/testing/testing.html). There is an example available in [EventConsumerTest](https://github.com/RedHatInsights/notifications-backend/blob/master/backend/src/test/java/com/redhat/cloud/notifications/events/EventConsumerTest.java#L42).

The idea is to replace the incoming Kafka `aggregation` channel with a mocked channel `InMemoryConnector.source("aggregation")` and then send a payload throught that channel. It will trigger a reactive messaging processing in `EmailSubscriptionTypeProcessor#consumeEmailAggregations` exactly like what would happen on prod. Then you should be able to run the Mockito `verify`.

Comment (bennetelli) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java:67]: I am not able to test it. Debugging doesn't work, changing the method to:     

@Incoming("aggregation")
    public Uni<Void> consumeEmailAggregations(String aggregationsJson) throws JsonProcessingException {throw new IllegalArgumentException("");}

Also doesn't result in exceptions in tests. I lost so many hours and days with this stuff and don't want to lose more. Testing single methods which deal with Unis and Multis doesn't work and testing this is also not fun. I will stop it here and mark it as resolved.

Comment (gwenneg) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java:67]: I can try fixing this if you want.

Comment (gwenneg) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java:90]: I pushed a test for the Kafka messages processing. This only works for successfull test cases.

Testing a failure because of an invalid payload will require another kind of test based on [this](https://smallrye.io/smallrye-mutiny/guides/testing).

Comment (bennetelli) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java:90]: Thanks :) I actually wanted to test more specific things, but maybe this is enough.

**Comment by pilhuhn** [comment]
Thank you @gwenneg for helping here.

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:364]: If we merge this, the old way of sending the daily email will be gone. That means the new way _has to_ work and be reliable immediately. Otherwise, we won't be able to fix any production issue that would require to change and redeploy notifications-backend.

Shouldn't we keep the old way and migrate to the new way using a configuration property?

cc @josejulio @pilhuhn

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/aggregators/PoliciesEmailPayloadAggregator.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/processors/email/aggregators/PoliciesEmailPayloadAggregator.java:42]: Unless I missed something, this is never used.

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/Email.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/processors/email/Email.java:9]: There are lots of access modifier changes and minor refactors in this PR which make the review longer/harder, especially in `EmailSubscriptionTypeProcessor`. Please don't put too many of these in a PR that contains a major change such as the daily email job. 😃

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:364]: I would like to put it on stage first and check if it's working there. 
There is no plan to put it in prod immediately. But this approach could be a problem when someone else also wants to put something in prod in the meantime.

By using a feature toggle, this would be a no brainer. 

I think we should keep the old way, indeed.

**Review by gwenneg — COMMENTED** [review]
Intermediate comments (I'm not done with the review):

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/Email.java
Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/Email.java:9]: Argh.. sorry. Got used to the pathfinder rule and clean up things the moment I see them.

File: backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailTest.java
Comment (gwenneg) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailTest.java:59]: It should also stay there for a while if we keep both the old and the new ways to aggregate emails in the code.

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:248]: Note for later: we'll need metrics in this method.

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:250]: If that line throws a `JsonProcessingException`, the Kafka message won't be acknowledged and there will be another attempt at processing the Kafka message which will likely be invalid again (there's no reason an invalid JSON payload would magically become valid). That sounds like an infinite loop of failed attempts.

If that happens, should we ack the Kafka message and add an ERROR log entry so that we get notified through Sentry that something went wrong and we have to manually fix it? If so, we may need to save the data that was unsuccessfully sent.

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:254]: Is there an NPE risk here because `emailAggregationItem.getPayload()` would be null?

**Review by gwenneg — CHANGES_REQUESTED** [review]
More comments, still not done 😃

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (josejulio) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:364]: Yes, lets keep the old way meanwhile and use a flag (maybe a variable in properties?) to turn on/off (default on :D).

File: backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java
Comment (josejulio) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java:62]: :cry:

Comment (bennetelli) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java:62]: Hey :-D the suggestion to use .await().indefinitely() comes from you :-D

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:254]: I would say no. There is a NotNull annotation on the payload attribute.

    @NotNull
    @Convert(converter = JsonObjectConverter.class)
    private JsonObject payload;

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:248]: I will create a Jira ticket for it so we don't lose this note

File: backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java
Comment (gwenneg) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java:62]: You're right, but it changed from OK to bad practice recently 😄 It's still OK with Quarkus 1 but I'm sure we'll migrate to Quarkus 2 soon so we'd better stop awaiting indefinitely ASAP.

Comment (bennetelli) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java:62]: a lot happening under the hood of Quarkus :-D where did you get this information from? :)

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:364]: Is "having it on status ON" a joke? Or do you want it to be on ON by default? :D

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java]: That's what I wanted to test. Doesn't work. Maybe someone else wants to give it a try?

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (josejulio) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:364]: I mean to have the "old way"  as default :D

Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:364]: Okay. I disabled the "new way" with my last commit. Your desired behavior should be there now

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java]: As far as I understand this here correctly: https://smallrye.io/smallrye-reactive-messaging/smallrye-reactive-messaging/2/acknowledgement/acknowledgement.html

They will get acknowledged with my latest change.

File: backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java
Comment (bennetelli) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java:66]: This would be a no brainer in "normal"  I am unable to test it in this reactive, kafka scenario.

File: backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java
Comment (bennetelli) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java:67]: I need help here. With Spring Boot this would be a no brainer. With this reactive and kafka thing it's not. At least not for me.

File: backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java
Comment (gwenneg) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java]: I've been involved in the Quarkus dev for more than 2 years so I naturally follow many subjects there. Being a liaison between our team and Quarkus is also part of my current responsibilites 😃

**Comment by gwenneg** [comment]
Please let me resolve the comments I created, otherwise I'm never sure which ones need my attention and I have to check all of them 😃

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:264]: We need the Sentry alert with a full stack trace if that happens:
```suggestion
            log.error("Kafka aggregation payload parsing failed", e);
```

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java]: The infinite loop is broken indeed. Now there's still the question of the data that may be lost after the scheduled aggregations purge following a failure here.

File: backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailTest.java
Comment (gwenneg) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailTest.java]: This class should be left untouched until `aggregator` is enabled by default on prod. We need that test coverage to prevent regressions in backend.

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:364]: The [old scheduled job](https://github.com/bennetelli/notifications-backend/blob/7ec7cfc6126588c19a68c9459947c605df33db32/backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java#L288-L292) should be executed only if `notifications.aggregator.email.subscription.periodic.cron.enabled` is `true`.

Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:364]: I am not sure if this would work. This entry is in another project and they are in different jars and have different application.properties. I would say this doesn't work.
We can add another property in notifcations-backend, but this somehow doesnt feel the right way

Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:364]: Oh yes I forgot notifications was multi-module now, my bad 😄

I think adding the same config property in backend is fine as long as both properties (backend and aggregator) are commented with something explaining that they should always be modified in a synchronized way and the values should never be identical between both projects.

File: backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailTest.java
Comment (bennetelli) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailTest.java]: added the missing tests again

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java]: Do you have an idea how to test it?

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:364]: I just added the property and comments in both application properties.

File: backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java
Comment (gwenneg) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java:68]: I pushed a small test that ensures that we won't throw any exception (in other words the Kafka message will be acked) in case of invalid payload.

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java]: See [this](https://github.com/RedHatInsights/notifications-backend/pull/379#discussion_r674636428) for the test.

I can only think of one case where we could have a payload parsing issue which could cause data loss: it could happen if `backend` and `aggregator` used different versions of the `EmailAggregation` class. As soon as both projects depend on the same version of that class from the `models` module, the problem will be gone.

The data loss isn't such a big deal, I don't think it's worth implementing anything to save that data in case of failure.

**Comment by bennetelli** [comment]
> Please let me resolve the comments I created, otherwise I'm never sure which ones need my attention and I have to check all of them smiley

Sorry.. :(

File: aggregator/src/main/resources/application.properties
Comment (gwenneg) [inline aggregator/src/main/resources/application.properties:44]: The scheduled job starts at 2 am in `backend` while it starts at 12 am here. Is there any reason for that change?

**Comment by gwenneg** [comment]
> Sorry.. :(

No worries 😉

File: backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java
Comment (gwenneg) [reply backend/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java:69]: Let's not block this PR because of that test, it can be added later.

**Comment by gwenneg** [comment]
I am [one comment](https://github.com/RedHatInsights/notifications-backend/pull/379#discussion_r674651589) away from approving the PR, but be careful with the merge as it will conflict pretty badly with https://github.com/RedHatInsights/notifications-backend/pull/388 (which I also reviewed). You should talk with @josejulio to determine who's the lucky guy who gets to merge first 😄

File: aggregator/src/main/resources/application.properties
Comment (bennetelli) [reply aggregator/src/main/resources/application.properties:44]: The property: notifications.backend.email.subscription.daily.cron=0 0 2 * * ? is the same as the whole time. I just renamed it.

The new property is also the same as before. I mean Thomas suggested that time. But anyway, we will change it anyway as soon as it's on ci to something like "every 1-2 minutes" or so to be able to test it.

I would say the current time in the new cron job is not that important for now as it will be changed at least two times anyway.

Comment (gwenneg) [reply aggregator/src/main/resources/application.properties:44]: I'm only concerned about prod:

- [Current prod cron value](https://github.com/RedHatInsights/notifications-backend/blob/56c53f81c8f24f0cf35f124298ab816879c359a5/backend/src/main/resources/application.properties#L93)  is `0 0 2 * * ?` (2 am)
- [Backend cron value after merge](https://github.com/bennetelli/notifications-backend/blob/22fc8a207ed4fe1b8a6d9bc62d528dc51ac44e70/backend/src/main/resources/application.properties#L102) will be `0 0 2 * * ?` (same as current prod)
- [Aggregator cron value after merge](https://github.com/bennetelli/notifications-backend/blob/22fc8a207ed4fe1b8a6d9bc62d528dc51ac44e70/aggregator/src/main/resources/application.properties#L44) will be `0 0 0 * * ?` (two hours earlier than current prod)

I would expect these to be all equal unless there's a good reason to change the execution hour.

Comment (gwenneg) [reply aggregator/src/main/resources/application.properties:44]: I didn't check if this is overridden using env vars on prod though.

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java]: > I am [one comment](https://github.com/RedHatInsights/notifications-backend/pull/379#discussion_r674651589) away from approving the PR, but be careful with the merge as it will conflict pretty badly with #388 (which I also reviewed). You should talk with @josejulio to determine who's the lucky guy who gets to merge first smile

Yea, I talked to him already :-D luckily he is still sleeping :-D so lets go :-D

Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java]: > I can only think of one case where we could have a payload parsing issue which could cause data loss: it could happen if `backend` and `aggregator` used different versions of the `EmailAggregation` class. As soon as both projects depend on the same version of that class from the `models` module, the problem will be gone.

I just tried to move all the classes to the models project. This doesn't work at the moment. We would need to add reactive things to the aggregator project first.

File: aggregator/src/main/resources/application.properties
Comment (bennetelli) [reply aggregator/src/main/resources/application.properties:44]: But aggregator won't get deployed to prod after merged this PR. There will be a second PR after we got it up and running on ci as I am sure there will be some things to clean up, maybe some things won't work, ... and so on. 
And we will also have to write the aggregator in reactive so we can get the benefits of the models project.

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java
Comment (bennetelli) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java]: > > I am [one comment](https://github.com/RedHatInsights/notifications-backend/pull/379#discussion_r674651589) away from approving the PR, but be careful with the merge as it will conflict pretty badly with #388 (which I also reviewed). You should talk with @josejulio to determine who's the lucky guy who gets to merge first smile
> 
> Yea, I talked to him already :-D luckily he is still sleeping :-D so lets go :-D

huhu, looking at you @josejulio :-D

File: aggregator/src/main/resources/application.properties
Comment (gwenneg) [reply aggregator/src/main/resources/application.properties:44]: Ok let's take care of that later then.

**Review by gwenneg — APPROVED** [review]
LGTM

**Comment by gwenneg** [comment]
There is one failing test! Will @josejulio win the merge race??

**Comment by josejulio** [comment]
> There is one failing test! Will @josejulio win the merge race??

I'm sure I can merge faster ;D

**Comment by gwenneg** [comment]
The test should be OK now.

**Comment by josejulio** [comment]
After syncing with @bennetelli we are going to merge this one; I'm going to assess the impact on the other PR and we are  going to check what would it take for the aggregator to pass the AggregationKey to kafka instead of the aggregated payload.

**Pull request merged**
